src/string.cr

```diff
@@ -3088,8 +3088,18 @@ class String
   # "abcdef".compare("ABCDEF", case_insensitive: true) == 0 # => true
   # ```
   def ==(other : self) : Bool
+    # Quick pointer comparison if both strings are identical references
     return true if same?(other)
+
+    # If the bytesize differs, they cannot be equal
     return false unless bytesize == other.bytesize
+
+    # If the character size of both strings differs, they cannot be equal.
+    # We need to exclude the case that @length of either string might not have
+    # been calculated (indicated by `0`).
+    return false unless (@length == other.@length) || (@length & other.@length).zero?
```

Comment:
To me it seems it would look nicer to un-invert the conditionals now.
"If sizes are not equal, we know the strings are not equal".

Also I'd suggest to not write this as a bitwise operation. I'm guessing that to the optimizer this can only make things less obvious.

```suggestion
    return false if bytesize != other.bytesize

    # If the character size of both strings differs, they cannot be equal.
    # We need to exclude the case that @length of either string might not have
    # been calculated (indicated by `0`).
    return false if @length != other.@length && @length != 0 && other.@length != 0
```

Reply:
Inverting the conditions sounds good 👍 

I figured the bitwise operation might be unnecessary and LLVM would optimize the zero comparisons anyway. But it actually seems to make a difference. The bitwise operation ends up using fewer instructions: https://godbolt.org/z/foY6j574K

```asm
equal_or_zero:
        cmpl    %esi, %edi
        sete    %al
        testl   %edi, %edi
        sete    %cl
        orb     %al, %cl
        testl   %esi, %esi
        sete    %al
        orb     %cl, %al
        retq

equal_or_zero_bitwise:
        cmpl    %esi, %edi
        sete    %cl
        testl   %edi, %esi
        sete    %al
        orb     %cl, %al
        retq
```

Reply:
`3 & 4 == 0`. I think a bitwise operation here is entirely incorrect

Reply:
🤦 Of course, thanks for pointing that out. Not sure why I thought that would be correct.
I should've treated the fact that LLVM wouldn't optimize the individual comparisons as a warning sign. If this could be optimized, LLVM would do it 🚀